### PR TITLE
error test failing (Not solution)

### PR DESCRIPTION
**Pull request by vladarskyi**

### CODE REVIEW

File: src/api/todos.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  return client.get<Todo[]>(`/todos?userId=${USER_ID}`);
+};
+
+export const createTodo = ({ title, userId, completed }: Omit<Todo, 'id'>) => {
+  return client.post<Todo>('/todos', {
+    userId,
+    title,
+    completed,
+  });
+};
```
Comment (Nirvacsh) [inline src/api/todos.ts:16]: You can shorten this to 
export const createTodo = (data: Omit<Todo, 'id'>) => {
  return client.post<Todo>('/todos', data);
};

**Review by Nirvacsh — COMMENTED** [review]
GJ, keep it up.

File: src/App.tsx [hunk 2]
```diff
@@ -1,26 +1,162 @@
...
+  if (filterBy !== FilterBy.All) {
+    filteredTodos = filteredTodos.filter(todo => {
+      switch (filterBy) {
+        case FilterBy.Active:
+          return !todo.completed;
+        case FilterBy.Completed:
+          return todo.completed;
+        default:
+          throw new Error('Unknown filter type');
+      }
```
Comment (Nirvacsh) [inline src/App.tsx:26]: You have a filter type - it is a bad practice to throw here in such way, better will be to implement such case in error type.

**Review by Nirvacsh — CHANGES_REQUESTED** [review]

**Comment by vladarskyi** [comment]
I have a lot of problems with the tests, but I can't find a solution (And I don't see the difference with the working version either)
errors 1
errors 2

File: src/App.tsx [hunk 3]
```diff
@@ -1,26 +1,205 @@
...
+import { TodoList } from './TodoList';
+import { Footer } from './Footer';
+import { Errors } from './types/Errors';
+import { FilterBy } from './types/FiilterBy';
+import { ErrorNotification } from './ErrorNotification';
+import { wait } from './utils/fetchClient';
+
+type FilterTheTodos = (todos: Todo[], filterBy: FilterBy) => Todo[];
+
+const getFilteredTodos: FilterTheTodos = (todos, filterBy) => {
```
Comment (BudnikOleksii) [inline src/App.tsx:15]: Move this function to utils/helpers file/folder

File: src/App.tsx [hunk 4]
```diff
@@ -1,26 +1,205 @@
...
+  const [errorMessage, setErrorMessage] = useState<Errors | null>(null);
+  const [filterBy, setFilterBy] = useState<FilterBy>(FilterBy.All);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isDisabled, setIsDisabled] = useState(false);
+  const [titleText, setTitleText] = useState('');
+  const [deletingIDs, setDeletingIDs] = useState<number[]>([]);
+
+  const handleClearingError = () => setErrorMessage(null);
+  const handleChangingFilterBy = (value: FilterBy) => setFilterBy(value);
+  const handleClearingCompletedTodos = () => {
```
Comment (BudnikOleksii) [inline src/App.tsx:47]: Add blank lines between handlers

File: src/App.tsx [hunk 5]
```diff
@@ -1,26 +1,205 @@
...
+    const idsToDelete: number[] = [];
+
+    todos.forEach(todo => {
+      if (todo.completed) {
+        idsToDelete.push(todo.id);
+      }
+    });
+
+    setDeletingIDs(idsToDelete);
```
Comment (BudnikOleksii) [inline src/App.tsx:56]:
```suggestion
    setDeletingIDs(prevIdsToDelete => [...prevIdsToDelete, ...idsToDelete]);
```
Add ids to existing ones in case you have some other loading states, and the same logic for removing

File: src/App.tsx [hunk 6]
```diff
@@ -1,26 +1,205 @@
...
+
+    Promise.all(idsToDelete.map(id => removeTodo(id)))
+      .then(() => {
+        setTodos(prevTodos => prevTodos.filter(todo => !todo.completed));
+      })
+      .catch(() => {
+        setErrorMessage(Errors.Delete);
+        wait(3000).then(() => setErrorMessage(null));
```
Comment (BudnikOleksii) [inline src/App.tsx:64]: Create a handler that sets the error message and then removes it in 3 seconds as you have to do it for all requests
```typescript
const handleError = (errorMessage: Errors) => {
  setErrorMessage(Errors.Delete);
  wait(3000).then(() => setErrorMessage(null));
}; 
```

File: src/App.tsx [hunk 7]
```diff
@@ -1,26 +1,205 @@
...
+
+    setIsDisabled(true);
+
+    setTempTodo({ id: 0, title, userId: USER_ID, completed: false });
+
+    createTodo({
+      title,
+      userId: USER_ID,
+      completed: false,
+    })
```
Comment (BudnikOleksii) [inline src/App.tsx:91]:
```suggestion
    const newTodo = {
      title,
      userId: USER_ID,
      completed: false,
    };
    
    setTempTodo({ id: 0, ...newTodo });

    createTodo(newTodo)
```
DRY

File: src/App.tsx [hunk 8]
```diff
@@ -1,26 +1,205 @@
...
+        setErrorMessage(Errors.Add);
+        setTempTodo(null);
+      })
+      .finally(() => {
+        setIsDisabled(false);
+      });
+  };
+
+  const deleteTodo = (id: number) => {
+    setDeletingIDs(curIDs => [...curIDs, id]);
```
Comment (BudnikOleksii) [inline src/App.tsx:107]: Do like this, in all places where you need to add new ids or remove them

File: src/App.tsx [hunk 9]
```diff
@@ -1,26 +1,205 @@
...
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <header className="todoapp__header">
+          {/* this button should have `active` class only if all todos are completed */}
```
Comment (BudnikOleksii) [inline src/App.tsx:149]: Remove all comments

File: src/App.tsx [hunk 10]
```diff
@@ -1,26 +1,205 @@
...
+          />
+
+          <form
+            onSubmit={() =>
+              handleSubmit({
+                title: titleText,
+                id: 0,
+                userId: 378,
+                completed: false,
+              })
```
Comment (BudnikOleksii) [inline src/App.tsx:163]: - You can get this object in your `handleSubmit` instead of passing it as an argument.
- And `378` is a magic value, create a constant for it or use if it's `USER_ID`

File: src/TodoListExamples.tsx
Comment (BudnikOleksii) [inline src/TodoListExamples.tsx:1]: Remove unnecessary file

**Review by BudnikOleksii — CHANGES_REQUESTED** [review]
Good work 👍 
Check the comments. Also. I can't create a new todo with your app, fix and test your solution
https://www.loom.com/share/fb4e2c7161d94f9f98b240b87a484424?sid=0add2aab-9473-4436-9d19-9d7a5ec302a6

File: src/App.tsx [hunk 11]
```diff
@@ -1,26 +1,167 @@
...
+
+  useEffect(() => {
+    if (!errorMessage) {
+      return;
+    }
+
+    const timerID = setTimeout(() => setErrorMessage(null), 3000);
+
+    return () => clearTimeout(timerID);
+  }, [errorMessage, isDisabled]);
```
Comment (BudnikOleksii) [inline src/App.tsx:109]: You can do it in `ErrorNotification` component

File: src/App.tsx [hunk 12]
```diff
@@ -1,26 +1,167 @@
...
-        </a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <header className="todoapp__header">
```
Comment (BudnikOleksii) [inline src/App.tsx:120]: Create a header component

File: src/TodoItem.tsx [hunk 13]
```diff
@@ -0,0 +1,32 @@
+import { Todo } from './types/Todo';
+
+interface Props {
+  tempTodo: Todo;
+}
+
+export const TodoItem: React.FC<Props> = ({ tempTodo }) => (
```
Comment (BudnikOleksii) [inline src/TodoItem.tsx:7]: You already have a todo component, use it

**Review by BudnikOleksii — CHANGES_REQUESTED** [review]
Almost done, get rid of the second todo component

**Review by BudnikOleksii — APPROVED** [review]
👍
